File: src/main/scala/laoflch/debezium/connector/informix/InformixStreamingChangeEventSource.scala
package laoflch.debezium.connector.informix

import com.informix.jdbc.IfmxReadableType
import io.debezium.data.Envelope.Operation
import io.debezium.pipeline.{ErrorHandler, EventDispatcher}
import io.debezium.pipeline.source.spi.{ChangeEventSource, StreamingChangeEventSource}
import io.debezium.relational.TableId
import io.debezium.util.Clock
import org.slf4j.{Logger, LoggerFactory}
import com.informix.stream.api.{IfmxStreamRecord, IfmxStreamRecordType}
import com.informix.stream.cdc.records.IfxCDCOperationRecord
import io.debezium.time.Timestamp

import scala.collection.mutable
import scala.jdk.javaapi.OptionConverters

object InformixStreamingChangeEventSource {


 /* var BEGIN =

  type COMMIT=COMMIT
  type DELETE=DELETE
  type DISCARD=DISCARD
  type ERROR=ERROR
  type INSERT=INSERT
  type METADATA=METADATA
  type ROLLBACK=ROLLBACK
  type TIMEOUT=TIMEOUT
  type TRUNCATE=TRUNCATE
  type BEFORE_UPDATE=BEFORE_UPDATE
  type AFTER_UPDATE=AFTER_UPDATE*/

}

class InformixStreamingChangeEventSource(connectorConfig: InformixConnectorConfig,
                                         offsetContext: InformixOffsetContext,
                                         dataConnection: InformixConnection,
                                         //metadataConnection: InformixConnection,
                                         dispatcher: EventDispatcher[TableId],
                                         errorHandler: ErrorHandler,
                                         clock: Clock,
                                         schema: InformixDatabaseSchema) extends StreamingChangeEventSource {


  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[InformixStreamingChangeEventSource])


  /**
   * Connection used for reading CDC tables.
   */
  //private val dataConnection: InformixConnection = dataConnection

  /**
   * A separate connection for retrieving timestamps; without it, adaptive
   * buffering will not work.
   */
  //private val metadataConnection: InformixConnection = metadataConnection

  //private val dispatcher: EventDispatcher[TableId] = dispatcher
  //private val errorHandler: ErrorHandler = errorHandler
  //private val clock: Clock = clock
  // private val schema: InformixDatabaseSchema = schema
  // private val offsetContext: InformixOffsetContext = offsetContext
  // private val pollInterval: Duration = connectorConfig.getPollInterval
  // private val connectorConfig: InformixConnectorConfig = connectorConfig


  @throws[InterruptedException]
  override def execute(context: ChangeEventSource.ChangeEventSourceContext): Unit = {
    val cdcEngine = dataConnection.getCDCEngine()
    val transactionCache = offsetContext.getInformixTransactionCache
    val schema = this.schema

    /*
    LOGGER.info("Schema          = {}", schema)
    LOGGER.info("Schema.tableIds = {}", schema.tableIds)
    CollectionConverters.SetHasAsScala(schema.tableIds()).asScala.foreach(x => {
      LOGGER.info("Schema Table: catalog={} schema={} table={}", x.catalog(), x.schema(), x.table())
      CollectionConverters.ListHasAsScala(schema.tableFor(x).columns()).asScala.foreach(c => {
        LOGGER.info("Schema Table Columns: name={} jdbcType={} nativeType={}, typeName={}",
          c.name(), c.jdbcType(), c.nativeType(), c.typeName())
      })
    })
    */

    /**
     * Initialize CDC Engine.
     */
    val lastPosition = offsetContext.getChangePosition
    val fromLsn: Long = lastPosition.getCommitLsn
    cdcEngine.setStartLsn(fromLsn)
    cdcEngine.init(schema)
    val label2TableId = cdcEngine.convertLabel2TableId()

    /**
     * Main Handle Loop
     */
    while (context.isRunning) {
      cdcEngine.stream((record) => {

        record.getType match {
          case IfmxStreamRecordType.TIMEOUT => {
            handleTimeOutEvent(record)
          }

          /**
           *
           * Handle BEFORE_UPDATE
           *
           * */
          case IfmxStreamRecordType.BEFORE_UPDATE => {

            val data = record.asInstanceOf[IfxCDCOperationRecord].getData

            offsetContext.setChangePosition(TxLogPosition.cloneAndSet(offsetContext.getChangePosition,
              TxLogPosition.LSN_NULL,   // commit
              record.getSequenceId,     // change
              record.getTransactionId,  // txnid
              TxLogPosition.LSN_NULL))  // begin

            transactionCache.beforeUpdate(record.getTransactionId, data)
          }

          /**
           *
           * Handle AFTER_UPDATE
           *
           * */
          case IfmxStreamRecordType.AFTER_UPDATE => {

            val newData = record.asInstanceOf[IfxCDCOperationRecord].getData
            val oldData = transactionCache.afterUpdate(record.getTransactionId).get

            offsetContext.setChangePosition(TxLogPosition.cloneAndSet(offsetContext.getChangePosition,
              TxLogPosition.LSN_NULL,
              record.getSequenceId,
              record.getTransactionId,
              TxLogPosition.LSN_NULL))

            val tableId = label2TableId(record.getLabel.toInt)

            handleEvent(tableId, offsetContext, InformixChangeRecordEmitter.OP_UPDATE, oldData, newData, clock, null)
          }

          /**
           *
           * Handle BEGIN
           *
           * */
          case IfmxStreamRecordType.BEGIN => {

            val _start = System.nanoTime()
            OptionConverters.toScala(transactionCache.beginTxn(record.getTransactionId)) match {
              case Some(value) => {
                println(value)
              }
              case None => {
                offsetContext.setChangePosition(TxLogPosition.cloneAndSet(offsetContext.getChangePosition,
                  TxLogPosition.LSN_NULL,
                  record.getSequenceId,
                  record.getTransactionId,
                  record.getSequenceId))

                offsetContext.getTransactionContext.beginTransaction(record.getTransactionId.toString)
              }
            }
            val _end = System.nanoTime()
            LOGGER.info("Received BEGIN :: transId={} seqId={} elaspedTs={}ms", record.getTransactionId(),  record.getSequenceId, (_end - _start)/1000000d)
          }

          /**
           *
           * Handle COMMIT
           *
           * */
          case IfmxStreamRecordType.COMMIT => {

            val _start = System.nanoTime()
            OptionConverters.toScala(transactionCache.commitTxn(record.getTransactionId)) match {
              case Some(value) => {
                offsetContext.setChangePosition(TxLogPosition.cloneAndSet(offsetContext.getChangePosition,
                  record.getSequenceId,
                  record.getSequenceId,
                  record.getTransactionId,
                  TxLogPosition.LSN_NULL))

                handleCommitEvent(offsetContext, value)
              }
              case None => {
                //val tableId=label2TableId(record.getLabel.toInt)
              }
            }
            offsetContext.getTransactionContext.endTransaction()
            val _end = System.nanoTime()

            LOGGER.info("Received COMMIT :: transId={} seqId={} elaspedTime={} ms", record.getTransactionId(),  record.getSequenceId, (_end - _start)/1000000d)
          }

          /**
           *
           * Handle ROLLBACK
           *
           * */
          case IfmxStreamRecordType.ROLLBACK => {

            val _start = System.nanoTime()
            OptionConverters.toScala(transactionCache.rollbackTxn(record.getTransactionId)) match {
              case Some(value) => {

                offsetContext.setChangePosition(TxLogPosition.cloneAndSet(offsetContext.getChangePosition,
                  TxLogPosition.LSN_NULL,
                  record.getSequenceId,
                  record.getTransactionId,
                  TxLogPosition.LSN_NULL))

                /**
                 * when rollbackTxn do noting handle but log the discarded records
                 */
                LOGGER.info("Rollback Txn:" + record.getTransactionId)
                handleRollbackEvent(offsetContext, value)
              }
              case None => {
                //val tableId=label2TableId(record.getLabel.toInt)
              }
            }
            offsetContext.getTransactionContext.endTransaction()
            val _end = System.nanoTime()

            LOGGER.info("Received ROLLBACK :: transId={} seqId={} elaspedTime={} ms", record.getTransactionId(),  record.getSequenceId, (_end - _start)/1000000d)
          }

          /**
           *
           * Handle INSERT
           *
           * */
          case IfmxStreamRecordType.INSERT => {

            val _start = System.nanoTime()

            val data = record.asInstanceOf[IfxCDCOperationRecord].getData

            offsetContext.setChangePosition(TxLogPosition.cloneAndSet(offsetContext.getChangePosition,
              TxLogPosition.LSN_NULL,
              record.getSequenceId,
              record.getTransactionId,
              TxLogPosition.LSN_NULL))

            val tableId = label2TableId(record.getLabel.toInt)
            if (LOGGER.isTraceEnabled()) {
              LOGGER.trace("Lookup label to tableId : {} -> {}", record.getLabel.toInt, tableId)
            }
            handleEvent(tableId, offsetContext, InformixChangeRecordEmitter.OP_INSERT, null, data, clock, null)
            val _end = System.nanoTime()

            LOGGER.info("Received INSERT :: transId={} seqId={} elaspedTime={} ms", record.getTransactionId(), record.getSequenceId(), (_end - _start)/1000000d)
          }

          /**
           *
           * Handle DELETE
           *
           * */
          case IfmxStreamRecordType.DELETE => {
            LOGGER.info("Received DELETE :: transId={} seqId={}", record.getTransactionId(), record.getSequenceId())

            val data = record.asInstanceOf[IfxCDCOperationRecord].getData

            offsetContext.setChangePosition(TxLogPosition.cloneAndSet(offsetContext.getChangePosition,
              TxLogPosition.LSN_NULL,
              record.getSequenceId,
              record.getTransactionId,
              TxLogPosition.LSN_NULL))

            val tableId = label2TableId(record.getLabel.toInt)
            handleEvent(tableId, offsetContext, InformixChangeRecordEmitter.OP_DELETE, data, null, clock, null)
          }

          case _ => {
            LOGGER.info("Handle unknown record-type = {}", record.getType.toString)
          }
        }

        def handleEvent(tableId: TableId,
                        offsetContext: InformixOffsetContext,
                        //txn:Long,
                        operation: Int,
                        data: java.util.Map[String, IfmxReadableType],
                        dataNext: java.util.Map[String, IfmxReadableType],
                        clock: Clock,
                        timestamp: Timestamp) {

          offsetContext.event(tableId, clock.currentTime())

          val cre = new InformixChangeRecordEmitter(offsetContext, operation,
            InformixChangeRecordEmitter.convertIfxData2Array(data),
            InformixChangeRecordEmitter.convertIfxData2Array(dataNext), clock)

          // add event in transcation
          offsetContext.getInformixTransactionCache.addEvent2Tx(tableId, cre, offsetContext.getChangePosition.getTxId)
        }

        def handleTimeOutEvent(record: IfmxStreamRecord): Unit = {
          offsetContext.setChangePosition(TxLogPosition.cloneAndSet(offsetContext.getChangePosition,
            TxLogPosition.LSN_NULL,
            record.getSequenceId,
            TxLogPosition.LSN_NULL,
            TxLogPosition.LSN_NULL))
        }

        def handleCommitEvent(offsetContext: InformixOffsetContext, cre: InformixTransactionCache.TransactionCacheBuffer): Unit = {
          try {
            LOGGER.info("Handle Commit {} Events", cre.size)
            cre.getTransactionCacheRecords.forEach(r => {
              dispatcher.dispatchDataChangeEvent(r.getTableId, r.getInformixChangeRecordEmitter)
            });
            /*
            cre.foreach(tuple => {
              dispatcher.dispatchDataChangeEvent(tuple._1, tuple._2)
            })
            */

          } catch {
            case e: Exception => LOGGER.info("HandleCommit got exception: {}", e.toString);
            //case _ => println("handleCommitEvent failed!")
          }
        }

        def handleRollbackEvent(offsetContext: InformixOffsetContext,  cre: InformixTransactionCache.TransactionCacheBuffer): Unit = {
          try {
            // cre.foreach(tuple => LOGGER.info("id:" + tuple._1 + ":" + "ChangeRecord:" + tuple._2.toString))
            LOGGER.info("Handle Rollback {} Events", cre.size)
            //dispatcher.dispatchDataChangeEvent(tableId,cre)
          } catch {
            case e: Exception => e.printStackTrace()
            //case _ => println("handleCommitEvent failed!")
          }
        }

        false
      })
    }
  }

  //override def toString: String = "ChangeTablePointer [changeTable=" + changeTable + ", resultSet=" + resultSet + ", completed=" + completed + ", currentChangePosition=" + currentChangePosition + "]"
}

File: src/main/java/laoflch/debezium/connector/informix/InformixStreamingChangeEventSource.java
/*
 * Copyright Debezium-Informix-Connector Authors.
 *
 * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 */

package laoflch.debezium.connector.informix;

import java.sql.SQLException;
import java.util.Map;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.informix.jdbc.IfmxReadableType;
import com.informix.stream.api.IfmxStreamRecord;
import com.informix.stream.cdc.records.IfxCDCBeginTransactionRecord;
import com.informix.stream.cdc.records.IfxCDCCommitTransactionRecord;
import com.informix.stream.cdc.records.IfxCDCMetaDataRecord;
import com.informix.stream.cdc.records.IfxCDCOperationRecord;
import com.informix.stream.cdc.records.IfxCDCRollbackTransactionRecord;
import com.informix.stream.cdc.records.IfxCDCTimeoutRecord;
import com.informix.stream.cdc.records.IfxCDCTruncateRecord;
import com.informix.stream.impl.IfxStreamException;

import io.debezium.pipeline.ErrorHandler;
import io.debezium.pipeline.EventDispatcher;
import io.debezium.pipeline.source.spi.StreamingChangeEventSource;
import io.debezium.relational.TableId;
import io.debezium.relational.TableSchema;
import io.debezium.util.Clock;

public class InformixStreamingChangeEventSource implements StreamingChangeEventSource<InformixOffsetContext> {

    private static Logger LOGGER = LoggerFactory.getLogger(InformixStreamingChangeEventSource.class);

    private final InformixConnectorConfig config;
    private final InformixConnection dataConnection;
    private final EventDispatcher<TableId> dispatcher;
    private final ErrorHandler errorHandler;
    private final Clock clock;
    private final InformixDatabaseSchema schema;

    public InformixStreamingChangeEventSource(InformixConnectorConfig connectorConfig,
                                              InformixConnection dataConnection,
                                              EventDispatcher<TableId> dispatcher,
                                              ErrorHandler errorHandler,
                                              Clock clock,
                                              InformixDatabaseSchema schema) {
        this.config = connectorConfig;
        this.dataConnection = dataConnection;
        this.dispatcher = dispatcher;
        this.errorHandler = errorHandler;
        this.clock = clock;
        this.schema = schema;
    }

    /**
     * Executes this source. Implementations should regularly check via the given context if they should stop. If that's
     * the case, they should abort their processing and perform any clean-up needed, such as rolling back pending
     * transactions, releasing locks etc.
     *
     * @param context contextual information for this source's execution
     * @return an indicator to the position at which the snapshot was taken
     * @throws InterruptedException in case the snapshot was aborted before completion
     */
    @Override
    public void execute(ChangeEventSourceContext context, InformixOffsetContext offsetContext) throws InterruptedException {
        InformixCDCEngine cdcEngine = dataConnection.getCdcEngine();
        InformixTransactionCache transCache = offsetContext.getInformixTransactionCache();

        /*
         * Initialize CDC Engine before main loop;
         */
        TxLogPosition lastPosition = offsetContext.getChangePosition();
        Long fromLsn = lastPosition.getCommitLsn();
        cdcEngine.setStartLsn(fromLsn);
        cdcEngine.init(schema);

        /*
         * Recover Stage. In this stage, we replay event from 'commitLsn' to 'changeLsn', and rebuild the transactionCache.
         */
        while (context.isRunning()) {
            if (lastPosition.getChangeLsn() <= lastPosition.getCommitLsn()) {
                LOGGER.info("Recover skipped, since changeLsn='{}' >= commitLsn='{}'",
                        lastPosition.getChangeLsn(), lastPosition.getCommitLsn());
                break;
            }

            try {
                IfmxStreamRecord record = cdcEngine.getCdcEngine().getRecord();
                if (record.getSequenceId() >= lastPosition.getChangeLsn()) {
                    LOGGER.info("Recover finished: from {} to {}, now Current seqId={}",
                            lastPosition.getCommitLsn(), lastPosition.getChangeLsn(), record.getSequenceId());
                    break;
                }
                switch (record.getType()) {
                    case TIMEOUT:
                        handleTimeout(offsetContext, (IfxCDCTimeoutRecord) record);
                        break;
                    case BEFORE_UPDATE:
                        handleBeforeUpdate(offsetContext, (IfxCDCOperationRecord) record, transCache, true);
                        break;
                    case AFTER_UPDATE:
                        handleAfterUpdate(cdcEngine, offsetContext, (IfxCDCOperationRecord) record, transCache, true);
                        break;
                    case BEGIN:
                        handleBegin(offsetContext, (IfxCDCBeginTransactionRecord) record, transCache, true);
                        break;
                    case INSERT:
                        handleInsert(cdcEngine, offsetContext, (IfxCDCOperationRecord) record, transCache, true);
                        break;
                    case COMMIT:
                        handleCommit(offsetContext, (IfxCDCCommitTransactionRecord) record, transCache, true);
                        break;
                    case ROLLBACK:
                        handleRollback(offsetContext, (IfxCDCRollbackTransactionRecord) record, transCache, false);
                        break;
                    case METADATA:
                        handleMetadata(cdcEngine, (IfxCDCMetaDataRecord) record);
                        break;
                    case TRUNCATE:
                        handleTruncate(cdcEngine, offsetContext, (IfxCDCTruncateRecord) record, transCache, true);
                        break;
                    case DELETE:
                        handleDelete(cdcEngine, offsetContext, (IfxCDCOperationRecord) record, transCache, true);
                        break;
                    default:
                        LOGGER.info("Handle unknown record-type = {}", record.getType());
                }
            }
            catch (SQLException e) {
                LOGGER.error("Caught SQLException", e);
                errorHandler.setProducerThrowable(e);
            }
            catch (IfxStreamException e) {
                LOGGER.error("Caught IfxStreamException", e);
                errorHandler.setProducerThrowable(e);
            }
        }

        /*
         * Main Handler Loop
         */
        try {
            while (context.isRunning()) {
                cdcEngine.stream((IfmxStreamRecord record) -> {
                    switch (record.getType()) {
                        case TIMEOUT:
                            handleTimeout(offsetContext, (IfxCDCTimeoutRecord) record);
                            break;
                        case BEFORE_UPDATE:
                            handleBeforeUpdate(offsetContext, (IfxCDCOperationRecord) record, transCache, false);
                            break;
                        case AFTER_UPDATE:
                            handleAfterUpdate(cdcEngine, offsetContext, (IfxCDCOperationRecord) record, transCache, false);
                            break;
                        case BEGIN:
                            handleBegin(offsetContext, (IfxCDCBeginTransactionRecord) record, transCache, false);
                            break;
                        case INSERT:
                            handleInsert(cdcEngine, offsetContext, (IfxCDCOperationRecord) record, transCache, false);
                            break;
                        case COMMIT:
                            handleCommit(offsetContext, (IfxCDCCommitTransactionRecord) record, transCache, false);
                            break;
                        case ROLLBACK:
                            handleRollback(offsetContext, (IfxCDCRollbackTransactionRecord) record, transCache, false);
                            break;
                        case METADATA:
                            handleMetadata(cdcEngine, (IfxCDCMetaDataRecord) record);
                            break;
                        case TRUNCATE:
                            handleTruncate(cdcEngine, offsetContext, (IfxCDCTruncateRecord) record, transCache, false);
                            break;
                        case DELETE:
                            handleDelete(cdcEngine, offsetContext, (IfxCDCOperationRecord) record, transCache, false);
                            break;
                        default:
                            LOGGER.info("Handle unknown record-type = {}", record.getType());
                    }

                    return false;
                });
            }
        }
        catch (SQLException e) {
            LOGGER.error("Caught SQLException", e);
            errorHandler.setProducerThrowable(e);
        }
        catch (IfxStreamException e) {
            LOGGER.error("Caught IfxStreamException", e);
            errorHandler.setProducerThrowable(e);
        }
        catch (Exception e) {
            LOGGER.error("Caught Unknown Exception", e);
            errorHandler.setProducerThrowable(e);
        }
        finally {
            cdcEngine.close();
        }
    }

    public void handleTimeout(InformixOffsetContext offsetContext, IfxCDCTimeoutRecord record) {
        offsetContext.setChangePosition(
                TxLogPosition.cloneAndSet(
                        offsetContext.getChangePosition(),
                        TxLogPosition.LSN_NULL,
                        record.getSequenceId(),
                        TxLogPosition.LSN_NULL,
                        TxLogPosition.LSN_NULL));
    }

    public void handleMetadata(InformixCDCEngine cdcEngine, IfxCDCMetaDataRecord record) {

        LOGGER.info("Received A Metadata: type={}, label={}, seqId={}",
                record.getType(), record.getLabel(), record.getSequenceId());

        /*
         * IfxCDCEngine engine = cdcEngine.getCdcEngine();
         * List<IfxCDCEngine.IfmxWatchedTable> watchedTables = engine.getBuilder().getWatchedTables();
         * List<IfxColumnInfo> cols = record.getColumns();
         * for (IfxColumnInfo cinfo : cols) {
         * LOGGER.info("ColumnInfo: colName={}, {}", cinfo.getColumnName(), cinfo.toString());
         * }
         * 
         * for (IfxCDCEngine.IfmxWatchedTable tbl : watchedTables) {
         * LOGGER.info("Engine Watched Table: label={}, tabName={}", tbl.getLabel(), tbl.getTableName());
         * }
         */
    }

    public void handleBeforeUpdate(InformixOffsetContext offsetContext, IfxCDCOperationRecord record, InformixTransactionCache transactionCache, boolean recover)
            throws IfxStreamException {

        Map<String, IfmxReadableType> data = record.getData();
        Long transId = (long) record.getTransactionId();
        transactionCache.beforeUpdate(transId, data);

        if (!recover) {
            offsetContext.setChangePosition(
                    TxLogPosition.cloneAndSet(
                            offsetContext.getChangePosition(),
                            TxLogPosition.LSN_NULL,
                            record.getSequenceId(),
                            transId,
                            TxLogPosition.LSN_NULL));
        }

    }

    public void handleAfterUpdate(InformixCDCEngine cdcEngine, InformixOffsetContext offsetContext, IfxCDCOperationRecord record,
                                  InformixTransactionCache transactionCache, boolean recover)
            throws IfxStreamException, SQLException {
        Long transId = (long) record.getTransactionId();

        Map<String, IfmxReadableType> newData = record.getData();
        Map<String, IfmxReadableType> oldData = transactionCache.afterUpdate(transId).get();

        Map<Integer, TableId> label2TableId = cdcEngine.convertLabel2TableId();
        TableId tid = label2TableId.get(Integer.parseInt(record.getLabel()));
        handleEvent(tid, offsetContext, transId, InformixChangeRecordEmitter.OP_UPDATE, oldData, newData, clock);

        if (!recover) {
            offsetContext.setChangePosition(
                    TxLogPosition.cloneAndSet(
                            offsetContext.getChangePosition(),
                            TxLogPosition.LSN_NULL,
                            record.getSequenceId(),
                            transId,
                            TxLogPosition.LSN_NULL));
        }
    }

    public void handleBegin(InformixOffsetContext offsetContext, IfxCDCBeginTransactionRecord record, InformixTransactionCache transactionCache, boolean recover)
            throws IfxStreamException {
        long _start = System.nanoTime();

        Long transId = (long) record.getTransactionId();
        Long beginTs = record.getTime();
        Long seqId = record.getSequenceId();

        Optional<InformixTransactionCache.TransactionCacheBuffer> transactionCacheBuffer = transactionCache.beginTxn(transId, beginTs, seqId);
        if (!recover) {
            Optional<InformixTransactionCache.TransactionCacheBuffer> minTransactionCache = transactionCache.getMinTransactionCache();
            Long minSeqId = minTransactionCache.isPresent() ? minTransactionCache.get().getBeginSeqId() : record.getSequenceId();

            if (!transactionCacheBuffer.isPresent()) {
                offsetContext.setChangePosition(
                        TxLogPosition.cloneAndSet(
                                offsetContext.getChangePosition(),
                                minSeqId,
                                record.getSequenceId(),
                                transId,
                                record.getSequenceId()));

                offsetContext.getTransactionContext().beginTransaction(String.valueOf(record.getTransactionId()));
            }
        }

        long _end = System.nanoTime();

        LOGGER.info("Received BEGIN :: transId={} seqId={} time={} userId={} elapsedTs={}ms",
                record.getTransactionId(), record.getSequenceId(),
                record.getTime(), record.getUserId(),
                (_end - _start) / 1000000d);
    }

    public void handleCommit(InformixOffsetContext offsetContext, IfxCDCCommitTransactionRecord record, InformixTransactionCache transactionCache, boolean recover)
            throws InterruptedException, IfxStreamException {
        long _start = System.nanoTime();
        Long transId = (long) record.getTransactionId();
        Long endTime = record.getTime();

        Optional<InformixTransactionCache.TransactionCacheBuffer> transactionCacheBuffer = transactionCache.commitTxn(transId, endTime);
        if (!recover) {
            Optional<InformixTransactionCache.TransactionCacheBuffer> minTransactionCache = transactionCache.getMinTransactionCache();
            Long minSeqId = minTransactionCache.isPresent() ? minTransactionCache.get().getBeginSeqId() : record.getSequenceId();

            if (transactionCacheBuffer.isPresent()) {
                offsetContext.setChangePosition(
                        TxLogPosition.cloneAndSet(
                                offsetContext.getChangePosition(),
                                minSeqId,
                                record.getSequenceId(),
                                transId,
                                TxLogPosition.LSN_NULL));

                for (InformixTransactionCache.TransactionCacheRecord r : transactionCacheBuffer.get().getTransactionCacheRecords()) {
                    dispatcher.dispatchDataChangeEvent(r.getTableId(), r.getInformixChangeRecordEmitter());
                }
                LOGGER.info("Handle Commit {} Events, transElapsedTime={}",
                        transactionCacheBuffer.get().size(), transactionCacheBuffer.get().getElapsed());
            }
            offsetContext.getTransactionContext().endTransaction();
        }

        long _end = System.nanoTime();
        LOGGER.info("Received COMMIT :: transId={} seqId={} time={} elapsedTime={} ms",
                record.getTransactionId(), record.getSequenceId(),
                record.getTime(),
                (_end - _start) / 1000000d);
    }

    public void handleInsert(InformixCDCEngine cdcEngine, InformixOffsetContext offsetContext, IfxCDCOperationRecord record, InformixTransactionCache transactionCache,
                             boolean recover)
            throws IfxStreamException, SQLException {
        long _start = System.nanoTime();
        Long transId = (long) record.getTransactionId();

        Optional<InformixTransactionCache.TransactionCacheBuffer> minTransactionCache = transactionCache.getMinTransactionCache();
        Long minSeqId = minTransactionCache.isPresent() ? minTransactionCache.get().getBeginSeqId() : record.getSequenceId();

        Map<String, IfmxReadableType> data = record.getData();
        Map<Integer, TableId> label2TableId = cdcEngine.convertLabel2TableId();
        TableId tid = label2TableId.get(Integer.parseInt(record.getLabel()));
        handleEvent(tid, offsetContext, transId, InformixChangeRecordEmitter.OP_INSERT, null, data, clock);

        if (!recover) {
            offsetContext.setChangePosition(
                    TxLogPosition.cloneAndSet(
                            offsetContext.getChangePosition(),
                            minSeqId,
                            record.getSequenceId(),
                            transId,
                            TxLogPosition.LSN_NULL));
        }

        long _end = System.nanoTime();
        LOGGER.info("Received INSERT :: transId={} seqId={} elapsedTime={} ms",
                record.getTransactionId(), record.getSequenceId(),
                (_end - _start) / 1000000d);
    }

    public void handleRollback(InformixOffsetContext offsetContext, IfxCDCRollbackTransactionRecord record, InformixTransactionCache transactionCache, boolean recover)
            throws IfxStreamException {
        long _start = System.nanoTime();
        Long transId = (long) record.getTransactionId();

        Optional<InformixTransactionCache.TransactionCacheBuffer> transactionCacheBuffer = transactionCache.rollbackTxn(transId);
        if (!recover) {
            Optional<InformixTransactionCache.TransactionCacheBuffer> minTransactionCache = transactionCache.getMinTransactionCache();
            Long minSeqId = minTransactionCache.isPresent() ? minTransactionCache.get().getBeginSeqId() : record.getSequenceId();

            if (minTransactionCache.isPresent()) {
                minSeqId = minTransactionCache.get().getBeginSeqId();
            }
            if (transactionCacheBuffer.isPresent()) {
                offsetContext.setChangePosition(
                        TxLogPosition.cloneAndSet(
                                offsetContext.getChangePosition(),
                                minSeqId,
                                record.getSequenceId(),
                                transId,
                                TxLogPosition.LSN_NULL));

                LOGGER.info("Rollback Txn: {}", record.getTransactionId());
            }
            offsetContext.getTransactionContext().endTransaction();
        }

        long _end = System.nanoTime();
        LOGGER.info("Received ROLLBACK :: transId={} seqId={} elapsedTime={} ms",
                record.getTransactionId(), record.getSequenceId(),
                (_end - _start) / 1000000d);
    }

    public void handleTruncate(InformixCDCEngine cdcEngine, InformixOffsetContext offsetContext, IfxCDCTruncateRecord record, InformixTransactionCache transactionCache,
                               boolean recover)
            throws IfxStreamException, SQLException {
        long _start = System.nanoTime();
        Long transId = (long) record.getTransactionId();

        Optional<InformixTransactionCache.TransactionCacheBuffer> minTransactionCache = transactionCache.getMinTransactionCache();
        Long minSeqId = minTransactionCache.isPresent() ? minTransactionCache.get().getBeginSeqId() : record.getSequenceId();

        Map<Integer, TableId> label2TableId = cdcEngine.convertLabel2TableId();
        TableId tid = label2TableId.get(record.getUserId());
        handleEvent(tid, offsetContext, transId, InformixChangeRecordEmitter.OP_TRUNCATE, null, null, clock);

        if (!recover) {
            offsetContext.setChangePosition(
                    TxLogPosition.cloneAndSet(
                            offsetContext.getChangePosition(),
                            minSeqId,
                            record.getSequenceId(),
                            transId,
                            TxLogPosition.LSN_NULL));
        }

        long _end = System.nanoTime();
        LOGGER.info("Received TRUNCATE :: transId={} seqId={} elapsedTime={} ms",
                record.getTransactionId(), record.getSequenceId(),
                (_end - _start) / 1000000d);
    }

    public void handleDelete(InformixCDCEngine cdcEngine, InformixOffsetContext offsetContext, IfxCDCOperationRecord record, InformixTransactionCache transactionCache,
                             boolean recover)
            throws IfxStreamException, SQLException {
        long _start = System.nanoTime();
        Long transId = (long) record.getTransactionId();

        Optional<InformixTransactionCache.TransactionCacheBuffer> minTransactionCache = transactionCache.getMinTransactionCache();
        Long minSeqId = minTransactionCache.isPresent() ? minTransactionCache.get().getBeginSeqId() : record.getSequenceId();

        Map<String, IfmxReadableType> data = record.getData();
        Map<Integer, TableId> label2TableId = cdcEngine.convertLabel2TableId();
        TableId tid = label2TableId.get(Integer.parseInt(record.getLabel()));
        handleEvent(tid, offsetContext, transId, InformixChangeRecordEmitter.OP_DELETE, data, null, clock);

        if (!recover) {
            offsetContext.setChangePosition(
                    TxLogPosition.cloneAndSet(
                            offsetContext.getChangePosition(),
                            minSeqId,
                            record.getSequenceId(),
                            transId,
                            TxLogPosition.LSN_NULL));
        }

        long _end = System.nanoTime();
        LOGGER.info("Received DELETE :: transId={} seqId={} elapsedTime={} ms",
                record.getTransactionId(), record.getSequenceId(),
                (_end - _start) / 1000000d);
    }

    public void handleEvent(TableId tableId,
                            InformixOffsetContext offsetContext,
                            Long txId,
                            Integer operation,
                            Map<String, IfmxReadableType> data,
                            Map<String, IfmxReadableType> dataNext,
                            Clock clock)
            throws SQLException {

        offsetContext.event(tableId, clock.currentTime());

        TableSchema tableSchema = schema.schemaFor(tableId);
        InformixChangeRecordEmitter informixChangeRecordEmitter = new InformixChangeRecordEmitter(offsetContext, operation,
                InformixChangeRecordEmitter.convertIfxData2Array(data, tableSchema),
                InformixChangeRecordEmitter.convertIfxData2Array(dataNext, tableSchema), clock);

        offsetContext.getInformixTransactionCache().addEvent2Tx(tableId, informixChangeRecordEmitter, txId);
    }
}
